### PR TITLE
[SRV6] Sonic-swss changes for SRV6

### PR DESCRIPTION
**Pull request by kperumalbfn**
SONiC swss support for SRV6
- RouteOrch changes to trigger SRV6 nexthops and update route entries
- SRV6Orch changes to create SRV6 nexthops, tunnel and SRV6 MY_SID_ENTRY objects
- CRMOrch changes to include SRV6 MY_SID

sonic-swss-common changes for SRV6 APPL_DB tables: https://github.com/Azure/sonic-swss-common/pull/538
sonic-utiliities changes for SRV6 CRM: https://github.com/Azure/sonic-utilities/pull/1883
sonic-buildimage changes for SRV6 CRM: https://github.com/Azure/sonic-buildimage/pull/9004

SRV6 HLD: https://github.com/Azure/SONiC/pull/795

### CODE REVIEW

**Comment by kperumalbfn** [comment]
sonic-swss srv6 test scripts will be pushed soon.

**Comment by kperumalbfn** [comment]
@caizhenghui-juniper Please review this PR and approve it.

**Comment by kperumalbfn** [comment]
Trigger

**Comment by kperumalbfn** [comment]
@prsunny @venkatmahalingam @caizhenghui-juniper All VS tests are passing except test_chassis_system_lag_id_allocator_table_full and test_chassis_system_lag_id_allocator_del_id

These tests will be bypassed after this commit - https://github.com/Azure/sonic-swss/pull/2004

File: orchagent/routeorch.cpp [hunk 1]
```diff
@@ -688,7 +704,36 @@ void RouteOrch::doTask(Consumer& consumer)
...
+
+                        for (uint32_t i = 1; i < srv6_segv.size(); i++)
+                        {
+                            if (ipv.empty()) {
+                                ip = "0.0.0.0";
+                                nhg_str += NHG_DELIMITER + ip;
+                            }
+                            else
+                            {
+                                nhg_str += NHG_DELIMITER + ipv[i];
```
[the remark this answers is not in the document]
Comment (kperumalbfn) [reply orchagent/routeorch.cpp:728]: @caizhenghui-juniper For SRV6, only srv6_segv and srv6_src are required. When we bring in FRR or other control plane, nexthop IP will be used.

File: orchagent/nexthopkey.h [hunk 2]
```diff
@@ -89,8 +107,12 @@ struct NextHopKey
         return str;
     }
 
-    const std::string to_string(bool overlay_nh) const
+    const std::string to_string(bool overlay_nh, bool srv6_nh) const
```
[the remark this answers is not in the document]
Comment (kperumalbfn) [reply orchagent/nexthopkey.h:110]: yeah, overlay_nh argument is unused here. We can check if we can remove that later.

File: orchagent/routeorch.cpp [hunk 3]
```diff
@@ -688,7 +704,36 @@ void RouteOrch::doTask(Consumer& consumer)
...
+                            }
+                            nhg_str += NH_DELIMITER + srv6_segv[i];
+                            nhg_str += NH_DELIMITER + srv6_src[i];
+                        }
+                        nhg = NextHopGroupKey(nhg_str, overlay_nh, srv6_nh);
+                        SWSS_LOG_NOTICE("SRV6 route with nhg %s", nhg.to_string().c_str());
+                    }
+                    else if (overlay_nh == false && srv6_nh == false)
```
[the remark this answers is not in the document]
Comment (kperumalbfn) [reply orchagent/routeorch.cpp:736]: yeah, extra condition to ignore both overlay_nh and srv6_nh

File: orchagent/nexthopkey.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (caizhenghui-juniper) [reply orchagent/nexthopkey.h:110]: agree, maybe using one argument to indicate extra_nh_type.

File: orchagent/routeorch.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (caizhenghui-juniper) [reply orchagent/routeorch.cpp:736]: srv6_nh == true is already checked in previous else if, no need to add srv6_nh == false here.

File: orchagent/routeorch.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (caizhenghui-juniper) [reply orchagent/routeorch.cpp:728]: e.g. when FRR comes in and set ip nhg as two NHS but srv6 sets 3 NHs, when i==2, there would be problem at this line, right? the number of NHs for IP is not tied/equal to the number of NHs for srv6. the logic is flawed. if you only thinking about exclusively for SRV6 NH only case, then do check to make sure ipv is empty.

File: tests/test_srv6.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+        # check if asic db entries are all restored
+        assert tunnel_entries == get_exist_entries(self.adb.db_connection, "ASIC_STATE:SAI_OBJECT_TYPE_TUNNEL")
+        assert nexthop_entries == get_exist_entries(self.adb.db_connection, "ASIC_STATE:SAI_OBJECT_TYPE_NEXT_HOP")
+        assert route_entries == get_exist_entries(self.adb.db_connection, "ASIC_STATE:SAI_OBJECT_TYPE_ROUTE_ENTRY")
+
+# Add Dummy always-pass test at end as workaroud
+# for issue when Flaky fail on final test it invokes module tear-down before retrying
+def test_nonflaky_dummy():
+    pass
```
Comment (caizhenghui-juniper) [inline tests/test_srv6.py:260]: should we add some negative test cases? e.g. invalid SRv6 sids setting?

File: orchagent/routeorch.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (kperumalbfn) [reply orchagent/routeorch.cpp:728]: @caizhenghui-juniper Good point. For SRV6, will check for if ipv is empty.

File: orchagent/routeorch.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (kperumalbfn) [reply orchagent/routeorch.cpp:736]: Yeah, srv6_nh condition check is redundant. Will remove it

[the remark this answers is not in the document]
Comment (kperumalbfn) [reply orchagent/routeorch.cpp]: Done

File: orchagent/routeorch.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (kperumalbfn) [reply orchagent/routeorch.cpp]: Done

File: tests/test_srv6.py (hunk 4 above)
Comment (kperumalbfn) [reply tests/test_srv6.py:260]: @caizhenghui-juniper Will add the negative tests in the next PR.

File: orchagent/routeorch.cpp [hunk 5]
```diff
@@ -688,6 +704,32 @@ void RouteOrch::doTask(Consumer& consumer)
...
+                            if (ipv.empty()) {
+                                ip = "0.0.0.0";
+                                nhg_str += NHG_DELIMITER + ip;
+                            }
+                            nhg_str += NH_DELIMITER + srv6_segv[i];
+                            nhg_str += NH_DELIMITER + srv6_src[i];
+                        }
+                        nhg = NextHopGroupKey(nhg_str, overlay_nh, srv6_nh);
+                        SWSS_LOG_NOTICE("SRV6 route with nhg %s", nhg.to_string().c_str());
+                    }
```
Comment (caizhenghui-juniper) [inline orchagent/routeorch.cpp:732]: I still don't like this block. 
Can be simplified as :
if ipv not empty then send error log and continue. 
Else 
for loop
nhg_str += "0.0.0.0" + .......
 
This should be much easier to read.

Comment (kperumalbfn) [reply orchagent/routeorch.cpp:729]: Thanks..Done

**Comment by kperumalbfn** [comment]
@caizhenghui-juniper All comments are addressed. Please approve the PR after review.

**Review by caizhenghui-juniper — APPROVED** [review]

**Comment by kperumalbfn** [comment]
Thanks for the approval @caizhenghui-juniper 

@prsunny @venkatmahalingam Please review and approve before 202111 branch pull.

File: orchagent/orchdaemon.cpp [hunk 6]
```diff
@@ -297,7 +306,8 @@ bool OrchDaemon::init()
      * when iterating ConsumerMap. This is ensured implicitly by the order of keys in ordered map.
      * For cases when Orch has to process tables in specific order, like PortsOrch during warm start, it has to override Orch::doTask()
      */
-    m_orchList = { gSwitchOrch, gCrmOrch, gPortsOrch, gBufferOrch, gIntfsOrch, gNeighOrch, gNhgOrch, gRouteOrch, copp_orch, qos_orch, wm_orch, policer_orch, tunnel_decap_orch, sflow_orch, debug_counter_orch, gMacsecOrch};
+
```
Comment (prsunny) [inline orchagent/orchdaemon.cpp:309]: Please remove this extra line and also move this to line 383 as the end.

File: orchagent/crmorch.cpp [hunk 7]
```diff
@@ -65,6 +67,8 @@ const map<CrmResourceType, uint32_t> crmResSaiAvailAttrMap =
     { CrmResourceType::CRM_DNAT_ENTRY, SAI_SWITCH_ATTR_AVAILABLE_DNAT_ENTRY },
     { CrmResourceType::CRM_MPLS_INSEG, SAI_OBJECT_TYPE_INSEG_ENTRY },
     { CrmResourceType::CRM_MPLS_NEXTHOP, SAI_SWITCH_ATTR_AVAILABLE_IPV4_NEXTHOP_ENTRY },
+    { CrmResourceType::CRM_SRV6_MY_SID_ENTRY, SAI_OBJECT_TYPE_MY_SID_ENTRY },
+    { CrmResourceType::CRM_SRV6_NEXTHOP, SAI_OBJECT_TYPE_NEXT_HOP },
```
Comment (prsunny) [inline orchagent/crmorch.cpp:71]: isn't this SAI_SWITCH_ATTR_AVAILABLE_IPV4_NEXTHOP_ENTRY. @smaheshm has recently fixed the MPLS_NEXTHOP above

File: orchagent/routeorch.cpp [hunk 8]
```diff
@@ -586,9 +590,17 @@ void RouteOrch::doTask(Consumer& consumer)
 
                     if (fvField(i) == "weight")
                         weights = fvValue(i);
-
+
```
Comment (prsunny) [inline orchagent/routeorch.cpp:593]: Please remove this tab space

File: orchagent/routeorch.cpp [hunk 9]
```diff
@@ -688,6 +704,29 @@ void RouteOrch::doTask(Consumer& consumer)
...
+                    {
+                        string ip;
+                        if (ipv.empty())
+                        {
+                            ip = "0.0.0.0";
+                        }
+                        else
+                        {
+                            SWSS_LOG_ERROR("For SRV6 nexthop ipv should be empty");
+                            continue;
```
Comment (prsunny) [inline orchagent/routeorch.cpp:717]: missing `it = consumer.m_toSync.erase(it);` ?

File: orchagent/routeorch.cpp [hunk 10]
```diff
@@ -893,6 +932,21 @@ void RouteOrch::doTask(Consumer& consumer)
             {
                 removeOverlayNextHops(it_nhg.second, it_nhg.first);
             }
+            else if (it_nhg.first.is_srv6_nexthop())
+            {
+                if(it_nhg.first.getSize() > 1)
+                {
+                    if(m_syncdNextHopGroups[it_nhg.first].ref_count == 0)
+                    {
+                      SWSS_LOG_NOTICE("Remove NHG in doTask");
```
Comment (prsunny) [inline orchagent/routeorch.cpp:941]: No information in the log. Please remove

File: orchagent/routeorch.cpp [hunk 11]
```diff
@@ -1612,19 +1686,28 @@ bool RouteOrch::addRoute(RouteBulkContext& ctx, const NextHopGroupKey &nextHops)
...
                     if (next_hop_id == SAI_NULL_OBJECT_ID)
                     {
-                        SWSS_LOG_ERROR("Failed to create Tunnel Nexthop %s", nexthop.to_string(overlay_nh).c_str());
+                        SWSS_LOG_ERROR("Failed to create Tunnel Nexthop %s", nexthop.to_string(overlay_nh, srv6_nh).c_str());
+                        return false;
+                    }
+                }
+                else if (srv6_nh)
+                {
+                    SWSS_LOG_NOTICE("Single NH: create srv6 nexthop %s", nextHops.to_string().c_str());
```
Comment (prsunny) [inline orchagent/routeorch.cpp:1707]: Please change the log level to INFO. Suggest not to have debug type logs in route add path. Revisit other 'NOTICE' logs as well in the PR

File: orchagent/srv6orch.cpp [hunk 12]
```diff
@@ -0,0 +1,654 @@
...
+    attr.id = SAI_NEXT_HOP_ATTR_TUNNEL_ID;
+    attr.value.oid = srv6_tunnel_id;
+    nh_attrs.push_back(attr);
+
+    status = sai_next_hop_api->create_next_hop(&nexthop_id, gSwitchId,
+                                                (uint32_t)nh_attrs.size(),
+                                                nh_attrs.data());
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("FAiled to create srv6 nexthop for %s", nh.to_string(false,true).c_str());
```
Comment (prsunny) [inline orchagent/srv6orch.cpp:198]: Failed. Fix the casing

File: orchagent/srv6orch.cpp [hunk 13]
```diff
@@ -0,0 +1,654 @@
...
+    set<NextHopKey> nexthops = nhgKey.getNextHops();
+    string srv6_source;
+    string srv6_segment;
+
+    for (auto nh : nexthops)
+    {
+        srv6_source = nh.srv6_source;
+        if (!createSrv6Tunnel(srv6_source))
+        {
+            SWSS_LOG_ERROR("Failed to create tunnel for source %s", srv6_source.c_str());
```
Comment (prsunny) [inline orchagent/srv6orch.cpp:220]: You want to continue create nexthop if Tunnel failed?

File: orchagent/orchdaemon.cpp (hunk 6 above)
Comment (kperumalbfn) [reply orchagent/orchdaemon.cpp]: Done

File: orchagent/crmorch.cpp (hunk 7 above)
Comment (kperumalbfn) [reply orchagent/crmorch.cpp]: Changed to SAI_SWITCH_ATTR_AVAILABLE_IPV6_NEXTHOP_ENTRY

File: orchagent/routeorch.cpp (hunk 8 above)
Comment (kperumalbfn) [reply orchagent/routeorch.cpp]: Done

File: orchagent/routeorch.cpp (hunk 10 above)
Comment (kperumalbfn) [reply orchagent/routeorch.cpp]: Removed the log

File: orchagent/routeorch.cpp (hunk 11 above)
Comment (kperumalbfn) [reply orchagent/routeorch.cpp]: Changed NOTICE logs to _INFO

File: orchagent/srv6orch.cpp (hunk 12 above)
Comment (kperumalbfn) [reply orchagent/srv6orch.cpp]: Done

File: orchagent/srv6orch.cpp (hunk 13 above)
Comment (kperumalbfn) [reply orchagent/srv6orch.cpp:220]: Thanks..done

File: orchagent/routeorch.cpp (hunk 9 above)
Comment (kperumalbfn) [reply orchagent/routeorch.cpp:718]: Done, for the invalid case.

File: orchagent/orchdaemon.cpp (hunk 6 above)
Comment (prsunny) [reply orchagent/orchdaemon.cpp:310]: i think you missed this comment

Comment (kperumalbfn) [reply orchagent/orchdaemon.cpp]: Thanks @prsunny. Removed the extra lines.

File: orchagent/nexthopgroupkey.h [hunk 14]
```diff
@@ -20,20 +21,37 @@ class NextHopGroupKey
...
+            m_srv6_nexthops = false;
+            auto nhv = tokenize(nexthops, NHG_DELIMITER);
+            for (const auto &nh_str : nhv)
+            {
+                auto nh = NextHopKey(nh_str, overlay_nh, srv6_nh);
+                m_nexthops.insert(nh);
+            }
+        }
+
+        if (srv6_nh)
```
Comment (venkatmahalingam) [inline orchagent/nexthopgroupkey.h:38]: srv6_nh and overlay_nh can co-exist?

File: orchagent/routeorch.cpp (hunk 5 above)
Comment (venkatmahalingam) [reply orchagent/routeorch.cpp:730]: why should nhg_str contain "0.0.0.0"?

File: orchagent/routeorch.cpp [hunk 15]
```diff
@@ -1612,19 +1686,28 @@ bool RouteOrch::addRoute(RouteBulkContext& ctx, const NextHopGroupKey &nextHops)
             /* IP neighbor is not yet resolved */
             else
             {
-                if(overlay_nh)
+                if(overlay_nh && !srv6_nh)
```
Comment (venkatmahalingam) [inline orchagent/routeorch.cpp:1689]: Both overlay_nh and srv6_nh can be true in any case?

Comment (kperumalbfn) [reply orchagent/routeorch.cpp:1689]: It can be either overlay_nh or srv6_nh, both cannot be true.

File: orchagent/nexthopgroupkey.h (hunk 14 above)
Comment (kperumalbfn) [reply orchagent/nexthopgroupkey.h:38]: cannot co-exist, will add else part.

Comment (kperumalbfn) [reply orchagent/nexthopgroupkey.h]: Done

**Review by prsunny — APPROVED** [review]

**Comment by prsunny** [comment]
@venkatmahalingam , @caizhenghui-juniper , please review/sign-off.

**Review by venkatmahalingam — APPROVED** [review]

**Comment by kperumalbfn** [comment]
Thanks @prsunny @venkatmahalingam 

@prsunny @caizhenghui-juniper's approval got cleared after code changes.

**Review by caizhenghui-juniper — APPROVED** [review]

**Comment by kperumalbfn** [comment]
@prsunny Please merge

**Comment by reshmaintel** [comment]
@prsunny @venkatmahalingam @caizhenghui-juniper @zhangyanzhao 
Thank you for your review help so far. Could you please help with the last step review/merge for this PR as well.

**Pull request merged**